orchagent/dash/dashpavalidationorch.cpp

```diff
@@ -0,0 +1,364 @@
+#include "dashpavalidationorch.h"
+
+#include "dashvnetorch.h"
+#include "crmorch.h"
+
+#include "taskworker.h"
+#include "pbutils.h"
+#include "saihelper.h"
+#include "swssnet.h"
+
+#include <cinttypes>
+#include <boost/functional/hash.hpp>
+
+extern sai_dash_pa_validation_api_t* sai_dash_pa_validation_api;
+extern size_t gMaxBulkSize;
+extern CrmOrch *gCrmOrch;
+extern sai_object_id_t gSwitchId;
+
+using namespace std;
+using namespace swss;
+
+size_t IpAddressHash::operator()(swss::IpAddress addr) const
+{
+    size_t seed = 0;
+    const auto &inner = addr.getIp();
+    boost::hash_combine(seed, inner.family);
+    if (inner.family == AF_INET)
+    {
+        boost::hash_combine(seed, inner.ip_addr.ipv4_addr);
+    }
+    else if (inner.family == AF_INET6)
+    {
+        boost::hash_combine(seed, inner.ip_addr.ipv6_addr);
+    }
+    return seed;
+}
+
+DashPaValidationOrch::DashPaValidationOrch(swss::DBConnector *db, swss::ZmqServer *zmqServer):
+    vnetOrch_(nullptr),
+    ZmqOrch(db, {APP_DASH_PA_VALIDATION_TABLE_NAME}, zmqServer)
+{
+    SWSS_LOG_ENTER();
+}
+
+void DashPaValidationOrch::setVnetOrch(const DashVnetOrch *vnetOrch)
+{
+    vnetOrch_ = vnetOrch;
+}
+
+bool DashPaValidationOrch::parseVni(const std::string& key, uint32_t& vni) {
+    std::istringstream iss(key);
+
+    iss >> vni;
+
+    return !iss.fail() && iss.eof();
+}
+
+bool DashPaValidationOrch::entryRefCountInc(const PaValidationEntry& entry)
+{
+    SWSS_LOG_ENTER();
+
+    bool new_entry = false;
+
+    auto vni = vni_map_.find(entry.vni);
+    if (vni == vni_map_.end())
+    {
+        new_entry = true;
+        vni_map_[entry.vni] = {{entry.address, 1}};
+    } else {
+        auto& vni_addresses = vni->second;
+        auto address = vni_addresses.find(entry.address);
+        if (address == vni_addresses.end())
+        {
+            new_entry = true;
+            vni_addresses[entry.address] = 1;
+        } else {
+            vni_addresses[entry.address]++;
```

Comment:
Can this condition happen? PA validation entry addition for same {VNI, IpAddress}??

Reply:
Yes, as part of DASH_VNET_MAPPING_TABLE processing:
https://github.com/sonic-net/sonic-swss/blob/1847195cf02a2baa478a664a671adc95c704a810/orchagent/dash/dashvnetorch.cpp#L373